Added NotesViewController, NotesViewModel, and Notes Storyboard

rmh211 opened the pull request:
## Overview of changes
Added add note button to add tree screen
Added NotesViewController, NotesViewModel, and Notes Storyboard
Added note saving ability to TreeService

## How was the change tested
- tested on device

## Screenshots (if applicable)

rmh211 committed:
Added NotesViewController, NotesViewModel, and Notes Storyboard
Added add note button to add tree screen
Added note saving ability to TreeService

rmh211 committed:
Working on adding memory for tree notes before saving tree

rmh211 committed:
Made sure note returned when saved and went to edit again before saving tree

rmh211 committed:
Made sure not to save placeholder text

rmh211 committed:
Change bundle identifier back

AlexCornforth commented on TreeTracker/UI/Views/Notes/NotesViewController.swift:
```diff
@@ -0,0 +1,78 @@
+    @IBOutlet private weak var notesTextField: UITextView! {
+        didSet {
+            notesTextField.layer.cornerRadius = 5
+            notesTextField.layer.borderColor = #colorLiteral(red: 0.5254901961, green: 0.7607843137, blue: 0.1960784314, alpha: 1)
```
There is a color (or colour if you speak proper english 😜) asset catalogue we can use for colors. Swiftgen will then generate color enums for them. `Asset.Colors` I believe.

There may even be a color already in there that is fit for purpose here.

AlexCornforth commented on TreeTracker/UI/Views/Notes/NotesViewController.swift:
```diff
@@ -0,0 +1,78 @@
+    @IBOutlet private weak var saveButton: PrimaryButton! {
+        didSet {
+            saveButton.isEnabled = true
+            saveButton.isHidden = false
```
Looks like we don't hide this at any point - would suggest we don't need this if so. Just ensure its not hidden in the storyboard.

AlexCornforth replied:
Likewise with saveButton.isEnabled

rmh211 replied:
I had considered it at one time, but hadn't removed them when I changed my mind. Thank you.

AlexCornforth commented on TreeTracker/UI/Views/Notes/NotesViewController.swift:
```diff
@@ -0,0 +1,78 @@
+
+extension NotesViewController: NotesViewModelViewDelegate {
+    func notesViewModel(_ notesViewModel: NotesViewModel, didHaveSavedNote note: String) {
+        print(note)
```
Would suggest removing print statements once they have served their purpose during development. If it is something we may want to log in the future we can use Logger in the core library which will print only in debug builds. But I'm not sure this is something we would want to log - it would be more for logging errors etc

AlexCornforth commented on TreeTracker/UI/Views/Notes/NotesViewModel.swift:
```diff
@@ -0,0 +1,37 @@
+//
+
+import UIKit
+import Treetracker_Core
```
Do we need to import this here? Can't see anywhere it is used.

rmh211 replied:
No, we don't need it here afterall

AlexCornforth commented on TreeTracker/UI/Views/Notes/NotesViewController.swift:
```diff
@@ -0,0 +1,78 @@
+            notesTextField.textColor = UIColor.lightGray
+            notesTextField.text = L10n.Notes.placeholder
+        }
+        let toolbar: UIToolbar = UIToolbar(frame: CGRect(x: 0, y: 0, width: self.view.frame.size.width, height: 30))
```
Maybe take a look at `KeyboardDismissing.swift` and how it is used. SignIn/SignUp are examples of using it to dismiss the keybaord.

rmh211 replied:
Originally, the textview took up so much of the screen, this protocol wasn't obvious to the user because you always tapped into the text view. I made the border more visible and moved the textview a little to make tapping outside of the textview more obvious. Now we don't need the done toolbar.

AlexCornforth replied:
Ah I understand - that makes sense. I think your addition of the done button was the right call then - sorry about that.

Probably worth keeping the background tap too though for consistent behaviour - although I imagine the button would get used the majority of the time.

AlexCornforth commented on TreeTracker/UI/Views/AddTree/AddTreeViewModel.swift:
```diff
@@ -32,17 +32,19 @@ class AddTreeViewModel {
         locationDataCapturer: LocationDataCapturing,
-        planter: Planter
+        planter: Planter,
+        note: String
```
Not sure we need to initialise with a note as we won't have one at this point. Unless I'm missing something.

rmh211 replied:
I now have note being passed around as an optional and only being unwrapped that the very end when it is needed. saveTree method can take an optional note because the TreeCapture has it as an optional property.

AlexCornforth reviewed (changes requested):
Looking great - nice work 👍 

Few suggestions - all pretty minor though.

rmh211 committed:
Keep note optional throughout (thus removing the need to initial a non existent note)
Removed Done toolbar option and replaced with KeyboardDismissing Protocol
Changed the note textview size and border to offer more obvious area for dismissing keyboard
Removed unused .isEnabled and .isHidden options
Textview Border color from color assets

rmh211 commented:
I've made all the suggested edits.

AlexCornforth commented on TreeTracker/UI/Navigation/Home/HomeCoordinator.swift:
```diff
@@ -107,7 +112,7 @@ private extension HomeCoordinator {
     }
 
-    func addTreeViewController(planter: Planter) -> UIViewController {
+    func addTreeViewController(planter: Planter, withNote note: String = "") -> UIViewController {
```
Looks like we can remove the note parameter here now

AlexCornforth reviewed (approved):
Great stuff Rob 🙏 

1 minor change if you wish but I'll leave it up to you as its not causing any issues. More than happy to merge this as it is 👍

rmh211 committed:
Added done button back to notes view
took unneeded parameter out.

AlexCornforth reviewed (approved):
🎉 🎉 🎉
